src/mozilla_sec_eia/ex_21/inference.py

```diff
@@ -0,0 +1,342 @@
+        predictions[np.where(inverse_indices == i)[0]] for i in range(len(unique_boxes))
+    ]
+    pred_counts = [np.bincount(arr) for arr in unique_box_predictions]
+    # Compute the mode of predictions for each group
+    # break ties by taking into account LABEL_PRIORITY
+    modes = np.array(
+        [
+            _sort_by_label_priority(np.where(arr == np.max(arr))[0])[0]
+            for arr in pred_counts
+        ]
+    )
+    flattened_modes = modes[inverse_indices]
+
+    return flattened_modes
+
+
+def _get_data(dataset):
+    yield from dataset
+
+
+def perform_inference(
+    pdfs_dir: Path,
+    model: LayoutLMv3ForTokenClassification,
+    processor: AutoProcessor,
+    dataset_ind: list = None,
+    labeled_json_dir: Path = None,
+    has_labels: bool = False,
+    device="cpu",
+):
+    """Predict entities with a fine-tuned model and extract Ex. 21 tables.
+
+    Arguments:
+        pdfs_dir: Path to the directory with PDFs that are being used for inference.
+        model: A fine-tuned LayoutLM model.
+        processor: The tokenizer and encoder for model inputs.
+        dataset_ind: A list of index numbers of dataset examples to be used for inference
+            Default is None, in which the entire dataset created from the PDF directory
+            is used.
+        labeled_json_dir: Path to the directory with labeled JSONs from Label Studio. Cannot
+            be None if has_labels is True.
+        has_labels: Boolean, true if the data has associated labels that can be used in
+            visualizing and validating results.
+        device: String or int, specify what computation device to use for inference
+            i.e. "mps", "cpu", "cuda"
+
+    Returns:
+        logits: The model outputs logits of shape (batch_size, seq_len, num_labels). Seq_len is
+            the same as token length
+        predictions: From the logits, we take the highest score for each token, using argmax.
+            This serves as the predicted label for each token. It is shape (seq_len) or token
+            length.
+        output_dfs: The extracted Ex. 21 tables.
+    """
+    dataset = create_inference_dataset(
+        pdfs_dir=pdfs_dir, labeled_json_dir=labeled_json_dir, has_labels=has_labels
+    )
+    if dataset_ind:
+        dataset = dataset.select(dataset_ind)
+
+    # TODO: figure out device argument
+    pipe = pipeline(
+        "token-classification",
+        model=model,
+        tokenizer=processor,
+        pipeline_class=LayoutLMInferencePipeline,
+        device=device,
+    )
+
+    logits = []
+    predictions = []
+    all_output_df = pd.DataFrame()
+    for logit, pred, output_df in pipe(_get_data(dataset)):
+        logits.append(logit)
+        predictions.append(pred)
+        all_output_df = pd.concat([all_output_df, output_df])
+    all_output_df = all_output_df[["id", "subsidiary", "loc", "own_per"]]
+    return logits, predictions, all_output_df
+
+
+class LayoutLMInferencePipeline(Pipeline):
```

Comment:
This is a Hugging Face `token-classification` pipeline that I made for inference and table extraction. Model inputs go first through `preprocess`, then those preprocessed outputs go into `_forward` and those outputs finally go through `postprocess` to get final outputs (logits, predictions, output dataframe). Within `postprocess` it calls `extract_table` to get the extracted Ex. 21 table. 

This model is supposed to make batch inference possible and generally speed things up but Hugging Face warns that if not implemented correctly then it can slow down inference. I need to do some memory profiling to make sure this works.